File: language_construct_performance/python/scripts/json_encoding.py
import os
import gc
import json
from benchmark_utils import benchmark_stats, save_stats

def create_plain_object(index):
    return {
        'label': f'user-dev-test-{index}',
        'id': f'u.{16406 + index}',
        'edgeTo': ['r.392', 'r.40', 'r.41', 'update', 'administrator', 'create', 'delete', 'read'],
        'accessTo': ['s.[s3].UACDR', 'a.[s3].DARC', 's.[secretsmanager].RACDU', 's.[dynamodb].RCDAU'],
        'details': {
            'provider': 'aws',
            'accountId': '568709751681',
            'principal': True,
            'tags': ['aKIAYI2NaRQPOT', 'dev testing local'],
            'mfas': '',
            'la': 1772454942 + (index % 1000),
            'ut': 2,
            's': 1,
            'cpd': 0,
            'pcb': '-',
            'lld': 0,
            'cd': 1763097939000,
            'cb': '-',
            'ub': '-',
            'ud': 0,
            'ua': 1772526871591,
            'ag': {
                's': {
                    't': 167
                },
                'a': {
                    't': 3187978,
                    's': {
                        't': 3187978,
                        's': 3149311,
                        'r': 42506
                    }
                }
            }
        }
    }

def run_benchmark():
    stats = {}
    columns = 3
    rows = 50000
    print(f"Starting JSON Encoding/Decoding Benchmark with {columns} columns x {rows} rows...\n")
    
    two_d_array = [None] * columns
    
    print("\n--- CREATION ---")
    def creation():
        for c in range(columns):
            column_array = [None] * rows
            for r in range(rows):
                column_array[r] = create_plain_object(c * rows + r)
            two_d_array[c] = column_array
    benchmark_stats("Creation", stats, "creationTimeMs", creation, track_memory=True)
    
    encoded_json = [None]
    
    print("\n--- JSON ENCODING ---")
    def json_encode():
        encoded_json[0] = json.dumps(two_d_array)
    benchmark_stats("JSON Encoding", stats, "jsonEncodeTimeMs", json_encode, track_memory=True)
    
    print("\n--- JSON DECODING ---")
    def json_decode():
        decoded = json.loads(encoded_json[0])
    benchmark_stats("JSON Decoding", stats, "jsonDecodeTimeMs", json_decode, track_memory=True)
    
    # Clear memory
    two_d_array.clear()
    encoded_json.clear()
    gc.collect()
    # Save Stats
    if not os.path.exists("data"):
        os.makedirs("data")
    json_stats = {
        "columns": columns,
        "rows": rows,
        "stats": stats
    }
    with open(os.path.join("data", "stats_json.json"), "w") as f:
        json.dump(json_stats, f, indent=2)
    print("\nSaved json stats to data/stats_json.json")

if __name__ == "__main__":
    try:
        run_benchmark()
    except Exception as e:
        print(f"Error: {e}")
        import sys
        sys.exit(1)

File: language_construct_performance/bun/scripts/idiomatic/json_encoding.ts
import { benchmarkStats, saveStats } from "./benchmark_utils";
import * as fs from "fs";

function createPlainObject(index: number) {
  return {
    label: "user-dev-test-" + index,
    id: "u." + (16406 + index),
    edgeTo: ["r.392", "r.40", "r.41", "update", "administrator", "create", "delete", "read"],
    accessTo: ["s.[s3].UACDR", "a.[s3].DARC", "s.[secretsmanager].RACDU", "s.[dynamodb].RCDAU"],
    details: {
      provider: "aws",
      accountId: "568709751681",
      principal: true,
      tags: ["aKIAYI2NaRQPOT", "dev testing local"],
      mfas: "",
      la: 1772454942 + (index % 1000),
      ut: 2,
      s: 1,
      cpd: 0,
      pcb: "-",
      lld: 0,
      cd: 1763097939000,
      cb: "-",
      ub: "-",
      ud: 0,
      ua: 1772526871591,
      ag: {
        s: {
          t: 167
        },
        a: {
          t: 3187978,
          s: {
            t: 3187978,
            s: 3149311,
            r: 42506
          }
        }
      }
    }
  };
}

const stats: Record<string, any> = {};

async function runBenchmark() {
  const columns = 3;
  const rows = 50000;
  console.log(`Starting JSON Encoding/Decoding Benchmark with ${columns} columns x ${rows} rows...`);

  let twoDArray = new Array(columns);

  console.log("\n--- CREATION ---");
  benchmarkStats("Creation", stats, "creationTimeMs", () => {
    for (let c = 0; c < columns; c++) {
      let columnArray = new Array(rows);
      for (let r = 0; r < rows; r++) {
        columnArray[r] = createPlainObject(c * rows + r);
      }
      twoDArray[c] = columnArray;
    }
  }, true);

  let encodedJSON: string;

  console.log("\n--- JSON ENCODING ---");
  benchmarkStats("JSON Encoding", stats, "jsonEncodeTimeMs", () => {
    encodedJSON = JSON.stringify(twoDArray);
  }, true);

  console.log("\n--- JSON DECODING ---");
  benchmarkStats("JSON Decoding", stats, "jsonDecodeTimeMs", () => {
    const decoded = JSON.parse(encodedJSON!);
  }, true);

  
  console.log("\n--- JSON FILE WRITE ---");
  benchmarkStats("JSON File Write", stats, "jsonFileWriteTimeMs", () => {
    fs.writeFileSync("data/test_dump.json", encodedJSON);
  }, true);

  let readJSON;
  console.log("\n--- JSON FILE READ ---");
  benchmarkStats("JSON File Read", stats, "jsonFileReadTimeMs", () => {
    readJSON = fs.readFileSync("data/test_dump.json", "utf-8");
  }, true);

  console.log("\n--- JSON FILE DECODE ---");
  benchmarkStats("JSON File Decode", stats, "jsonFileDecodeTimeMs", () => {
    const decoded = JSON.parse(readJSON);
  }, true);

    if (fs.existsSync("data/test_dump.json")) fs.unlinkSync("data/test_dump.json");
  // Clear memory
  twoDArray = null as any;
  encodedJSON = null as any;
  Bun.gc(true);

  // Save Stats
  import("fs").then(fs => {
    import("path").then(path => {
      if (!fs.existsSync("data")) fs.mkdirSync("data");
      const statsPath = "data/stats_json.json";
      let jsonStats = { columns, rows, stats: {} };
      if (fs.existsSync(statsPath)) {
        try { jsonStats = JSON.parse(fs.readFileSync(statsPath, "utf-8")); } catch(e) {}
      }
      jsonStats["idiomatic"] = stats;
      jsonStats.columns = columns;
      jsonStats.rows = rows;
      fs.writeFileSync(statsPath, JSON.stringify(jsonStats, null, 2));
      console.log("\nSaved json stats to data/stats_json.json");
    });
  });
}

runBenchmark().catch(console.error);
